### PR TITLE
Adding tutorials for deployment on Heroku and Digital Ocean

### PR DESCRIPTION
**Pull request by malloxpb**
Fixes: #

Improves: Since it took me a lot of time figuring out the platforms I should use to deploy my Plone site as well as how to deploy it, I made this tutorial to help people save time.

Changes proposed in this pull request:

-Guide users to deploying Plone on Heroku step-by-step to get deeper understanding on how it works

-Add the deployment on Digital Ocean tutorial to guide new users deploy their first Plone app on VPS more easily.

-Give common errors people might get into while deploying their apps on Digital Ocean.
[preview.pdf](https://github.com/plone/documentation/files/938859/preview.pdf)

### CODE REVIEW

File: manage/deploying/production/heroku_do.rst [hunk 1]
```diff
@@ -0,0 +1,226 @@
+=============================
+Deploy your Plone app to Heroku and Digital Ocean
+=============================
+
```
Comment (svx) [inline manage/deploying/production/heroku_do.rst]: Not valid reST, please follow: https://docs.plone.org/about/rst-styleguide.html

File: manage/deploying/production/heroku_do.rst [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+------------
+
+There are many ways to deploy a Plone site. 
+However, we sometimes would want to deploy our Plone sites for testing and developing Plone before we can find services that fit our needs.
+Therefore, this tutorial will show how to deploy Plone on Heroku and Digital Ocean, which are popular platforms for development.
+
+
+Deploying your Plone site on Heroku
+------------
+
```
Comment (svx) [inline manage/deploying/production/heroku_do.rst]: Not valid reST, please follow: https://docs.plone.org/about/rst-styleguide.html

File: manage/deploying/production/heroku_do.rst [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+=============================
+
+.. admonition:: Description
+
+    Since deploying a Plone site might be a challenge for new Plone users, this tutorial aims for solving those difficulties.
+
+
+Introduction
+------------
+
```
Comment (svx) [inline manage/deploying/production/heroku_do.rst:12]: Wring header, please follow: https://docs.plone.org/about/rst-styleguide.html

File: manage/deploying/production/heroku_do.rst [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+Heroku is very useful in deploying testing sites since it is free and flexible. 
+It supports many programming languages and necessary addons for different development environments.
+Deploying apps on Heroku is one of the way to show people our work samples without spending money.
+Plone has a `Heroku buildpack <https://plone.org/download>`_ that supports deploying automatically.
+On the other hand, this tutorial will show step-by-step deploying Plone app on Heroku manually for better understanding.
+
+Setting up Heroku account and local environment
+^^^^^^^^^
+
```
Comment (svx) [inline manage/deploying/production/heroku_do.rst]: Not valid reST, please follow: https://docs.plone.org/about/rst-styleguide.html

File: manage/deploying/production/heroku_do.rst [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+.. image:: ../images/herokuapp.png
+    :align: center
+    :alt: Heroku sign up page
+
+We then need to install Heroku Toolbelt, which is a tool that helps deploying apps from our local machine to Heroku.
+You can find an installer that suits your computer with full instruction at `Heroku documentation <https://devcenter.heroku.com/articles/heroku-cli>`_.
+
+Download and deploy Plone
+^^^^^^^^^
+
```
Comment (svx) [inline manage/deploying/production/heroku_do.rst]: Not valid reST and wrong header, please follow: https://docs.plone.org/about/rst-styleguide.html

File: manage/deploying/production/heroku_do.rst [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+We are going to use the training-sandbox Plone package because it is designed for Heroku deployment.
+Firstly, we will clone the Plone training-sandbox.
+
+.. code-block:: shell
+
+    git clone https://github.com/collective/training-sandbox
+    $ cd training-sandbox
+
```
Comment (svx) [inline manage/deploying/production/heroku_do.rst]: Please follow our style-guide: https://docs.plone.org/about/rst-styleguide.html

File: manage/deploying/production/heroku_do.rst [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+We then need to create a Heroku app.
+
+.. code-block:: shell
+
+    heroku create --buildpack https://github.com/plone/heroku-buildpack-plone [your-app-name]
+    $ git remote -v
+        heroku	https://git.heroku.com/[your-app-name].git (fetch)
+        heroku	https://git.heroku.com/[your-app-name].git (push)
+
```
Comment (svx) [inline manage/deploying/production/heroku_do.rst]: Please follow our style-guide: https://docs.plone.org/about/rst-styleguide.html

File: manage/deploying/production/heroku_do.rst [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+    heroku addons:create heroku-postgresql:hobby-dev --app [your-app-name]
+
+Now we should be good to go for deploying our app to Heroku
+
+.. code-block:: shell
+
+    git add .
+    $ git commit -m "your commit message"
+    $ git push heroku master
+
```
Comment (svx) [inline manage/deploying/production/heroku_do.rst]: Please follow our style-guide: https://docs.plone.org/about/rst-styleguide.html

File: manage/deploying/production/heroku_do.rst [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+
+Deploying your Plone app to Digital Ocean
+------------
+
```
Comment (svx) [inline manage/deploying/production/heroku_do.rst]: Please follow our style-guide: https://docs.plone.org/about/rst-styleguide.html

File: manage/deploying/production/heroku_do.rst [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+Digital Ocean is a service that provides VPS and root access for users, 
+which makes it very convenient since there are services that do not give full control over the VPS.
+We will use this platform for our tutorial since Digital Ocean is cheap and reliable.
+
+Setting up accounts and droplets
+^^^^^^^^^
+
```
Comment (svx) [inline manage/deploying/production/heroku_do.rst]: Please follow our style-guide: https://docs.plone.org/about/rst-styleguide.html

File: manage/deploying/production/heroku_do.rst [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+using our email and email password to sign up.
+
+After having your account ready, we will need to create a droplet.
+A droplet is the way Digital Ocean calls its own Virtual Private Server (VPS). 
+Click on the Create Droplet button on the right corner of the screen.
+
+.. image:: ../images/create_droplet.png
+    :align: center
+    :alt: Create droplet
+
```
Comment (svx) [inline manage/deploying/production/heroku_do.rst:34]: Wrong location for pictures, they should go into the _static under /root of the repo, 
please follow our style-guide: https://docs.plone.org/about/rst-styleguide.html

File: manage/deploying/production/heroku_do.rst [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+We will be redicrect to the droplet selection page. 
+In here, we need to check the OS for your Droplet. In this tutorial, we choose the Ubuntu 16.04 OS for our VPS.
+
+.. image:: ../images/create_dropletos.png
+    :align: center
+    :alt: Choose droplet OS
+
```
Comment (svx) [inline manage/deploying/production/heroku_do.rst:41]: Wrong location for pictures, they should go into the _static under /root of the repo,
please follow our style-guide: https://docs.plone.org/about/rst-styleguide.html

File: manage/deploying/production/heroku_do.rst [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+After that, we need to choose the plan for your Droplet. 
+Usually, we will use the $10/month since it provides good space for caching.
+In addition, this plan also prevents some errors that we might get into while deploying our Plone sites.
+
+.. image:: ../images/create_dropletsize.png
+    :align: center
+    :alt: Choose droplet size
+
```
Comment (svx) [inline manage/deploying/production/heroku_do.rst:49]: Wrong location for pictures, they should go into the _static under /root of the repo,
please follow our style-guide: https://docs.plone.org/about/rst-styleguide.html

File: manage/deploying/production/heroku_do.rst [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+.. note::
+    In case you want to deploy your app using a specific domain name, you can learn how to set it up at `Digital Ocean documentation <https://www.digitalocean.com/community/tutorials/how-to-set-up-a-host-name-with-digitalocean>`_
+
```
Comment (svx) [inline manage/deploying/production/heroku_do.rst:52]: Typo . is missing

File: manage/deploying/production/heroku_do.rst [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+Sign in to your Droplet 
+^^^^^^^^^
+
```
Comment (svx) [inline manage/deploying/production/heroku_do.rst]: Please follow our style-guide: https://docs.plone.org/about/rst-styleguide.html

File: manage/deploying/production/heroku_do.rst [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+Mac and Linux
+`````````
+
```
Comment (svx) [inline manage/deploying/production/heroku_do.rst]: Please follow our style-guide: https://docs.plone.org/about/rst-styleguide.html

File: manage/deploying/production/heroku_do.rst [hunk 17]
```diff
@@ -0,0 +1,226 @@
...
+When you are done with it, make sure that you are in the /root directory. Install the Unified Installer 
+
+.. code-block:: shell
+
+    wget --no-check-certificate https://launchpad.net/plone/5.0/5.0.7/+download/Plone-5.0.7-UnifiedInstaller.tgz
+
+Windows
+`````````
+
```
Comment (svx) [inline manage/deploying/production/heroku_do.rst]: Please follow our style-guide: https://docs.plone.org/about/rst-styleguide.html

File: manage/deploying/production/heroku_do.rst [hunk 18]
```diff
@@ -0,0 +1,226 @@
...
+On Windows, you will have to download `Putty <http://www.chiark.greenend.org.uk/~sgtatham/putty/latest.html>`_.
+Run the setup. When you are done, start Putty.
+Use your droplet ip address, port 22 and hit enter. It should open a linux terminal for you to fill in your username and password.
+Log in to your droplet with the password from the email that you received.
+
+.. image:: ../images/winputty.png
+    :align: center
+    :alt: Windows putty setup
+
```
Comment (svx) [inline manage/deploying/production/heroku_do.rst:86]: Wrong location for pictutres

File: manage/deploying/production/heroku_do.rst [hunk 19]
```diff
@@ -0,0 +1,226 @@
...
+When you are done, it shold ask you to change the password.
+
+Deploy Plone to VPS
+^^^^^^^^^
+
```
Comment (svx) [inline manage/deploying/production/heroku_do.rst]: Not valid reST and wrong header

File: manage/deploying/production/heroku_do.rst [hunk 20]
```diff
@@ -0,0 +1,226 @@
...
+    wget --no-check-certificate https://launchpad.net/plone/5.0/5.0.7/+download/Plone-5.0.7-UnifiedInstaller.tgz
+    $ tar -xf Plone-5.0.7-UnifiedInstaller.tgz
+
+Run the the Unified Installer
+
+.. code-block:: shell
+
+    cd Plone-5.0.7-UnifiedInstaller
+    $ ./install.sh $OPTION
+
```
Comment (svx) [inline manage/deploying/production/heroku_do.rst:105]: This is creating 'double' content, better use :ref: or :doc: for linking to the installation documentation

File: manage/deploying/production/heroku_do.rst [hunk 21]
```diff
@@ -0,0 +1,226 @@
...
+    For deployment on Digital Ocean, you should run the Installer with default installation path, which is /opt/plone
+
+After the installer has successfully installed Plone on our VPS, we will run the buildout and test the server.
+
+.. code-block:: shell
+
+    cd /opt/plone/zinstance
+    $ sudo -u plone_buildout bin/buildout
+    $ ./bin/instance fg
+
```
Comment (svx) [inline manage/deploying/production/heroku_do.rst:116]: Please follow our style-guide

File: manage/deploying/production/heroku_do.rst [hunk 22]
```diff
@@ -0,0 +1,226 @@
...
+follow from step 3 :doc:`Ubuntu Production deployment </manage/deploying/production/ubuntu_production>`.
+
+.. note::
+    You will have to replace the path specified in step 3 with your Plone app path. In this case, it should be /opt/plone/zinstance
+
+When you finish step 5, you should have your Plone app running at [your-host-name].com.
+
+Common errors
+^^^^^^^^^
+
```
Comment (svx) [inline manage/deploying/production/heroku_do.rst]: Wrong header and not valid reST

File: manage/deploying/production/heroku_do.rst [hunk 23]
```diff
@@ -0,0 +1,226 @@
...
+Cannot install lxml==3.5.0
+`````````
+
```
Comment (svx) [inline manage/deploying/production/heroku_do.rst]: Not valid reST

File: manage/deploying/production/heroku_do.rst [hunk 24]
```diff
@@ -0,0 +1,226 @@
...
+    Building without Cython.
+    Using build configuration of libxslt 1.1.28
+    Compile failed: command 'x86_64-linux-gnu-gcc' failed with exit status 4
+    error: Setup script exited with error: command 'x86_64-linux-gnu-gcc' failed with exit status 4
+    An error occurred when trying to install lxml 3.5.0. Look above this message for any errors that were output by easy_install.
+    While:
+      Installing instance.
+      Getting distribution for 'lxml==3.5.0'.
+    Error: Couldn't install: lxml 3.5.0
+
```
Comment (svx) [inline manage/deploying/production/heroku_do.rst:152]: Formating is not valid

File: manage/deploying/production/heroku_do.rst [hunk 25]
```diff
@@ -0,0 +1,226 @@
...
+This error will show up when you use the Unified Installer on the $5/month plan droplet. 
+You will need to have the $10/month in order to run the Unified Installer. 
+After finishing the setup, you can resize the droplet to 5$/month plan if you want.
+
+The Installer does not generate all the Plone files
+`````````
+.. image:: ../images/errorDO.png
```
Comment (svx) [inline manage/deploying/production/heroku_do.rst]: Not valid reST and wrong header

File: manage/deploying/production/heroku_do.rst [hunk 26]
```diff
@@ -0,0 +1,226 @@
...
+    :align: center
+    :alt: droplet installation error
+
+This error happens when the default installation path in Unified Installer has been modified.
+**How to fix**: run the Unified Installer again. But you need to leave the installation path as default.
```
Comment (svx) [inline manage/deploying/production/heroku_do.rst:164]: Using ..note:: would be better

File: manage/deploying/production/index.rst [hunk 27]
```diff
@@ -13,5 +13,6 @@ Production
 
     restarts
     ubuntu_production
+    heroku_do
     securitybestpractices
```
Comment (svx) [inline manage/deploying/production/index.rst:17]: Better use a filename which is clear indicating what it is about

**Comment by svx** [comment]
Thanks for you first contribution and taking the time to working on that !

There are a couple of 'issues' :)

First, please make sure that you follow [our style-guide](https://docs.plone.org/about/rst-styleguide.html), currently the reST syntax is  not valid for headers, code-examples and so on.

Second you use the wrong header types  :)

Third, we have one place where we store docs, this is the *_static* directory in the root.
This is also written in our style-guide :)

One other thing is that you creating double content, because you copy already written content and place it under a new location, it is better to link or reference here, this is easier to maintain and also better for user, because with your solution, this content shows up  twice now, if people searching for it, which is confusing :)

Taking is further, we have already docs about Heroku, there are located in the repositories:

- https://github.com/plone/heroku-buildpack-plone
- https://github.com/plone/heroku-button-plone

Following our setup, I would prefer to check if your parts can added there as improvements and than, if we really want, we fetch the docs from there.

About the DO docs, IMHO, you find on DO self, a excellent documentation, how you create droplets use ssh and stuff like this.
If we add all this to our docs, this creates extra work for us, because we have to make sure that this is always is sync and up to date with DO.

Currently at PLOG [Plone Open Garden] people are busy with working on a new installer overview site, IMHO this should be the place where we may add instructions, our docs are more about Plone itself :)

Never the less, thanks for all your work, this was for sure not for nothing, we 'just' have to see how and where we use it :)

Thank you again !

**Comment by stevepiercy** [comment]
In addition to the above suggestions, I found the content to be both unique and applicable to Plone from [line 151](https://github.com/nctl144/documentation/blame/c37cd8fb986793d9aa2a41c5268ea378706fc8f1/manage/deploying/production/heroku_do.rst#L151) onward. I doubt I would have figured out how to fix the lxml install issue.

**Comment by malloxpb** [comment]
Hey guys,

Thank you so much for your comments on this PR. I really appreciate it :). I will try to work on fixing the content as soon as possible and come up with solutions on how to use this in the Plone documentation :)

**Comment by malloxpb** [comment]
@svx After a few days working on this, I have deleted the Heroku part in this PR and I think I am going to contribute to the Heroku README files to make the deployment tutorial easier (if it is a good idea in your opinion). 
About the indentation and the ReST styling for the documentation, I have fixed the headings, images locations, etc.. However, I can see that the default branch has changed to 5.1. Should I make another pull request to that branch and close this PR?

**Pull request closed without merging**